Applications/Spire/Source/Ui/DropDownMenu2.cpp

```diff
@@ -0,0 +1,195 @@
+#include "Spire/Ui/DropDownMenu2.hpp"
+#include <QFocusEvent>
+#include <QScrollBar>
+#include <QVBoxLayout>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/ScrollArea.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+
+namespace {
+  auto BORDER_PADDING() {
+    static auto padding = 2 * scale_width(1);
+    return padding;
+  }
+
+  void adjust_current_index(DropDownMenu2& menu, int amount) {
+    if(menu.get_count() == 0 || amount == 0) {
+      return;
+    }
+    auto last_index = menu.get_count() - 1;
+    auto index = [&] {
+      auto index = menu.get_current();
+      if(!index) {
+        if(amount > 0) {
+          return 0;
+        } else {
+          return last_index;
+        }
+      }
+      *index += amount;
+      if(index > last_index) {
+        return 0;
+      } else if(index < 0) {
+        return last_index;
+      }
+      return *index;
+    }();
+    menu.set_current(index);
+  }
+}
+
+DropDownMenu2::DropDownMenu2(std::vector<DropDownMenuItem2*> items,
+    QWidget* parent)
+    : QWidget(parent, Qt::Tool | Qt::FramelessWindowHint),
+      m_selected_index(0),
+      m_max_displayed_items(5) {
+  setAttribute(Qt::WA_ShowWithoutActivating);
+  m_shadow = new DropShadow(true, false, this);
+  setObjectName("DropDownMenu");
+  setStyleSheet(R"(
+    #DropDownMenu {
+      background-color: #C8C8C8;
+    })");
+  auto layout = new QVBoxLayout(this);
+  layout->setContentsMargins(scale_width(1), 0, scale_width(1),
+    scale_height(1));
+  m_scroll_area = new ScrollArea(this);
+  m_scroll_area->setFocusProxy(parent);
+  m_scroll_area->setWidgetResizable(true);
+  layout->addWidget(m_scroll_area);
+  auto body = new QWidget(m_scroll_area);
+  m_list_layout = new QVBoxLayout(body);
+  m_list_layout->setContentsMargins({});
+  m_list_layout->setSpacing(0);
+  m_scroll_area->setWidget(body);
+  for(auto& item : items) {
+    m_list_layout->addWidget(item);
+    m_item_hovered_connections.AddConnection(item->connect_hovered_signal(
+      [=] { m_hovered_signal(item->get_value()); }));
+    m_item_selected_connections.AddConnection(item->connect_selected_signal(
+      [=] { on_item_selected(item->get_value(), item); }));
+  }
+  if(m_list_layout->count() > 0) {
+    update_height();
+    static_cast<DropDownMenuItem2*>(
+      m_list_layout->itemAt(0)->widget())->set_highlighted();
+    m_selected_index = 0;
+    m_current_index = 0;
+    m_selected_signal(get_value(*m_selected_index));
```

Comment:
These signals aren't doing anything, it's not possible for a connection to exist in the constructor.